File: nim/grade-school/Iteration1/grade_school.nim

import algorithm

type
  
  Student* = tuple [
    name: string,
    grade: int
  ]
  
  School* = object
    students*: seq[Student]

func compareStudents(a, b: Student): int =
  result = cmp(a.grade, b.grade)
  if result == 0:
    result = cmp(a.name, b.name)

func roster*(school: School): seq[string] =
  var s = school.students
  sort(s, compareStudents)

  for student in s:
    result.add(student.name)

func grade*(school: School, target: int): seq[string] =
  for student in school.students:
    if student.grade == target:
      result.add(student.name)
      sort(result, system.cmp[string])

File: fsharp/grade-school/Iteration1/GradeSchool.fs
module GradeSchool

type School = Map<int, string list>

let empty: School =
    Map.empty

let roster (school: School): string list =
    school
    |> Map.toList
    |> List.sortBy (fun kvp -> fst kvp)
    |> List.collect (fun kvp -> snd kvp |> List.sort)

let grade (grade: int) (school: School): string list =
    let maybeStudents =
        school
        |> Map.tryFind grade

    match maybeStudents with
    | Some students -> students |> List.sort
    | None -> []

let add (student: string) (grade: int) (school: School): School =
    let studentEnrolled = 
        roster school
        |> List.contains student

    match studentEnrolled with
    | true -> school
    | false ->
        school
        |> Map.change grade (fun g ->
            match g with
            | Some students -> Some (student :: students |> List.distinct)
            | None -> Some [student]
        )
